graspologic/embed/n2v.py

```diff
@@ -0,0 +1,482 @@
+
+    Further details on the Alias Method used in this functionality can be found at
+        https://lips.cs.princeton.edu/the-alias-method-efficient-sampling-with-many-discrete-outcomes/
+
+    References
+    ----------
+    .. [1] Aditya Grover and Jure Leskovec  "node2vec: Scalable Feature Learning for Networks."
+        Knowledge Discovery and Data Mining, 2016.
+    """
+
+    _preconditions(
+        graph,
+        num_walks,
+        walk_length,
+        return_hyperparameter,
+        inout_hyperparameter,
+        dimensions,
+        window_size,
+        workers,
+        iterations,
+        interpolate_walk_lengths_by_node_degree,
+    )
+
+    node2vec_graph = _Node2VecGraph(graph, return_hyperparameter, inout_hyperparameter)
+
+    logging.info(
+        f"Starting preprocessing of transition probabilities on graph with {str(len(graph.nodes()))} nodes and "
+        f"{str(len(graph.edges()))} edges"
+    )
+
+    start = time.time()
+    logging.info(f"Starting at time {str(start)}")
+
+    node2vec_graph._preprocess_transition_probabilities()
+
+    logging.info(f"Simulating walks on graph at time {str(time.time())}")
+    walks = node2vec_graph._simulate_walks(
+        num_walks, walk_length, interpolate_walk_lengths_by_node_degree
+    )
+
+    logging.info(f"Learning embeddings at time {str(time.time())}")
+    model = _learn_embeddings(walks, dimensions, window_size, workers, iterations)
+
+    end = time.time()
+    logging.info(
+        f"Completed. Ending time is {str(end)} Elapsed time is {str(start - end)}"
+    )
+
+    return model.wv.vectors, model.wv.index2word
+
+
+def _assert_is_positive_int(name: str, value: int):
+    if not isinstance(value, int):
+        raise TypeError(f"{name} must be an int")
+    if value <= 0:
+        raise ValueError(f"{name} must be > 0")
+
+
+def _assert_is_nonnegative_float(name: str, value: float):
+    if not isinstance(value, float):
+        raise TypeError(f"{name} must be a float")
+    if value < 0.0:
+        raise ValueError(f"{name} must be >= 0.0")
+
+
+def _preconditions(
+    graph: Union[nx.Graph, nx.DiGraph],
+    num_walks: int,
+    walk_length: int,
+    return_hyperparameter: float,
+    inout_hyperparameter: float,
+    dimensions: int,
+    window_size: int,
+    workers: int,
+    iterations: int,
+    interpolate_walk_lengths_by_node_degree: bool,
+):
+    if not isinstance(graph, nx.Graph):
+        raise TypeError("graph must be a networkx Graph or DiGraph")
+    if graph.is_multigraph():
+        raise ValueError(
+            "This function does not work on multigraphs - because there are two reasonable ways to treat a "
+            "multigraph with different behaviors, we insist that the caller create an appropriate Graph or "
+            "DiGraph that represents the manner in which they'd like the multigraph to be treated for the "
+            "purposes of this embedding"
+        )
+    _assert_is_positive_int("num_walks", num_walks)
```

Comment:
If we are going to standardize on having precondition checks then it is probably worthwhile to move our common assertions into a centralized location (similarly to how we implemented it in topologic.)

Reply:
I agree

Reply:
https://github.com/microsoft/graspologic/issues/558